hypothesis-python/src/hypothesis/database.py

```diff
@@ -324,3 +335,287 @@ def delete(self, key: bytes, value: bytes) -> None:
+            "This database is read-only. "
+            "Please wrap this class with ReadOnlyDatabase"
+            "i.e. ReadOnlyDatabase(GitHubArtifactsDatabase(...))."
+        )
+
+    def __repr__(self) -> str:
+        return f"GitHubArtifactDatabase(owner={self.owner}, repo={self.repo}, artifact_name={self.artifact_name})"
+
+    def _initialize_io(self) -> None:
+        if self._initialized:
+            return
+
+        # Load the artifact into memory
+        # This root allows us to access the files in the zipfile as if they were on the FS
+        # It's compatible with the pathlib Path API
+        self._initialized = True
+
+    def _initialize_db(self) -> None:
+        # Create the cache directory if it doesn't exist
+        mkdir_p(str(self.path))
+
+        # Get all artifacts
+        cached_artifacts = sorted(
+            self.path.glob("*.zip"),
+            key=lambda a: datetime.datetime.fromisoformat(a.stem),
+        )
+
+        # Remove all but the latest artifact
+        for new_artifact in cached_artifacts[:-1]:
+            new_artifact.unlink()
+
+        try:
+            found_artifact = cached_artifacts[-1]
+        except IndexError:
+            found_artifact = None
+
+        # Check if the latest artifact is a cache hit
+        if found_artifact is not None and (
+            datetime.datetime.now()
+            - datetime.datetime.fromisoformat(found_artifact.stem)
+            < self.cache_timeout
+        ):
+            self._artifact = found_artifact
+            self._initialize_io()
+            return
+
+        # pragma: no branch
+        # Download the latest artifact from GitHub
+        new_artifact = self._fetch_artifact()
+
+        if new_artifact:
+            if found_artifact is not None:
+                found_artifact.unlink()
+            self._artifact = new_artifact
+        elif found_artifact is not None:
+            warnings.warn(
+                HypothesisWarning(
+                    "Using an expired artifact as a fallback for the database."
+                )
+            )
+            self._artifact = found_artifact
+        else:
+            warnings.warn(HypothesisWarning("Disabling shared database due to errors."))
+            self._disabled = True
+            return
+
+        self._initialize_io()
+
+    def _fetch_artifact(self) -> Optional[Path]:  # pragma: no cover
+        # Get the list of artifacts from GitHub
+        request = Request(
+            f"https://api.github.com/repos/{self.owner}/{self.repo}/actions/artifacts",
+            headers={
+                "Accept": "application/vnd.github+json",
+                "X-GitHub-Api-Version": "2022-11-28 ",
+                "Authorization": f"Bearer {self.token}",
+            },
+        )
+
+        try:
+            response = urlopen(request)
+            artifacts = json.loads(response.read())["artifacts"]
+            warning_message = None
+        except HTTPError as e:
```

Comment:
Let's pull out this request-to-bytes-with-warnings code to a helper function?  It looks like we have the same messages twice.